matrix_sdk/src/client.rs

```diff
@@ -478,6 +478,66 @@ impl Client {
         Ok(())
     }
 
+    /// Gets the mxc avatar url of the owner of the client, if set.
+    ///
+    /// # Example
+    /// ```no_run
+    /// # use futures::executor::block_on;
+    /// # use matrix_sdk::Client;
+    /// # use url::Url;
+    /// # let homeserver = Url::parse("http://example.com").unwrap();
+    /// # block_on(async {
+    /// # let user = "example";
+    /// let client = Client::new(homeserver).unwrap();
+    /// client.login(user, "password", None, None).await.unwrap();
+    ///
+    /// if let Some(url) = client.avatar_url().await.unwrap() {
+    ///     println!("Your avatar's mxc url is {}", url);
+    /// }
+    /// ```
+    pub async fn avatar_url(&self) -> Result<Option<String>> {
+        let user_id = self.user_id().await.ok_or(Error::AuthenticationRequired)?;
+        let request = get_avatar_url::Request::new(&user_id);
+        let response = self.send(request).await?;
+        Ok(response.avatar_url)
+    }
+
+    /// Sets the mxc avatar url of the client's owner. The avatar gets unset if `url` is `None`.
+    pub async fn set_avatar_url(&self, url: Option<&str>) -> Result<()> {
+        let user_id = self.user_id().await.ok_or(Error::AuthenticationRequired)?;
+        let request = set_avatar_url::Request::new(&user_id, url);
+        self.send(request).await?;
+        Ok(())
+    }
+
+    /// Upload and set the owning client's avatar.
+    ///
+    /// The will upload the data produced by the reader to the homeserver's content repository, and
+    /// set the user's avatar to the mxc url for the uploaded file.
+    ///
+    /// This is a convenience method for calling [`upload()`](#method.upload), followed by
+    /// [`set_avatar_url()`](#method.set_avatar_url).
+    ///
+    /// # Example
+    /// ```no_run
+    /// # use std::{path::Path, fs::File, io::Read};
+    /// # use matrix_sdk::Client;
+    /// # block_on(async {
+    /// # let homeserver = Url::parse("http://locahost:8080").unwrap();
+    /// # let client = Client::new(homeserver).unwrap();
+    /// let path = Path::from("/home/example/selfie.jpg");
+    /// let mut image = File::open(path).unwrap();
+    ///
+    /// client.upload_avatar(&mime::IMAGE_JPG, &mut image).await.expect("Can't set avatar");
+    /// #})
+    /// ```
+    pub async fn upload_avatar<R: Read>(&self, content_type: &Mime, reader: &mut R) -> Result<()> {
```

Comment:
Can the avatar URL be of any mime type? Perhaps we should restrict this to image formats or perhaps restricting anything here is anyways futile. Any thoughts on this?

Reply:
I looked through the client-server spec and couldn't find any mentioned of requirements for mime types of avatars, so i figure it shouldn't be the sdk's role to restrict those